### ISSUE TITLE
Add @id to datatypes and use those in conversions

### ISSUE THREAD

**Issue by JeniT** [w3c/csvw#543]
This followed from discussion about #524. For example, this would mean I could do:

```
"datatype": {
  "@id": "http://example.org/datatypes/age",
  "base": "integer",
  "minimum": 0,
  "maximum": 120
}
```

The identifier `http://example.org/datatypes/age` would be irrelevant for validation purposes, but on conversion to RDF (or potentially XML), the identified datatype could be associated with the value. It could also offer a location for further information or definition of the datatype.

**Comment by gkellogg** [comment]
How about adding:

> Processors **MAY** use the value of `@id` to perform additional type checking.

Or something like that?

**Comment by iherman** [comment]
If we do that, this may be used for other things than XSD Schemas. OWL 2 has introduced the possibility to define what is called [Datatype Restrictions](http://www.w3.org/TR/2012/REC-owl2-syntax-20121211/#Datatype_Restrictions): essentially one can define in OWL2 (ie, in RDF) the datatype in the example. It would actually be much more natural to use that than the XML Schema datatypes for RDF purposes.

I am _not_ proposing that an RDF conversion would do this by default. But what this means is that if there is a note in the metadata document on the usage of that `@id` property than we should not give only XSD as an example (making it clear that the MAY above would be properly understood as performing additional checking possibly on OWL and not on XSD).

**Comment by gkellogg** [comment]
From the meeting:

[07:08am] danbri: gkellogg: I believe all the work is done. One concern ...
[07:08am] danbri: … if we allow a datatype that has ONLY the id, … which someone (Robert?) suggested -
[07:09am] danbri: … if we do that, from a json-ld perspective it is v similar to having a datatype which i believe has @type @id
[07:09am] danbri: … in which case it can take a value like an id
[07:09am] danbri: so it is like having a datatype with http://example.org/datatype/age
[07:09am] danbri: whereas we elsewhere say value of dt as a string becomes the base
[07:10am] danbri: ivan: I might have a def of a datatype, written down in schema eg owl, ...
[07:10am] danbri: let's not try to say that this is an id
[07:12am] jtandy: gkellogg: [this] is one of the things that I hate to bring up - pedantic
[07:12am] jtandy: ivan: no -it's ok
[07:13am] jtandy: ... the concern is that you're not identifying the object itself, you're identifying an external definition with is applied to the object
[07:13am] jtandy: ... so let's not use @id
[07:13am] jtandy: ... let's come up with a property name
[07:14am] jtandy: ivan: once this is done @jtandy or I need to make a change to the CSV2RDF doc
[07:14am] jtandy: ... and add a new property to the diagram & data model
[07:15am] jtandy: ivan: can you @gkellogg do the data model and metadata doc
[07:15am] jtandy: ... then ping @jtandy to make change in the CSV2RDF
[07:15am] danbri: scribenick: danbri
[07:15am] jtandy: https://github.com/w3c/csvw/issues/524
[07:15am] danbri: jtandy: at the bottom of issue #524
[07:16am] danbri: … the original about the xml datatypes, i suggested that with his example, with @id pointing to datatype age he might want to add a base statement, non-negative integer
[07:16am] danbri: …as you'd be informing the csvw validator that the no. in there should be a non-negative integer
[07:16am] danbri: … but then he responded saying validation is important
[07:17am] danbri: ivan: we do not require a validator to do that
[07:17am] danbri: … validation is nice to have but not req'd for conformance
[07:17am] danbri: he seemed to be ok for that
[07:17am] danbri: gkellogg: if a base for a dt doesn't exist, it is assumed to be string
[07:17am] jtandy: example: "datatype" : { "base": "nonNegativeInteger", "@id" : "http://example.org/datatypes/age" }
[07:17am] danbri: we might consider if we add a new property, … for base(?)
[07:17am] • danbri missed detail
[07:18am] danbri: gkellogg: reason for use of base was that it describes the base datatype that we're annotating
[07:18am] jtandy: [base property is used by the validators to check the value of the cell against a given lexical form]
[07:18am] danbri: … creates an annotated subset of that datatype
[07:18am] danbri: … just not defining rules that are for validation, transformation etc.
[07:19am] danbri: ...
[07:20am] danbri: [discussion of response workflow]
[07:20am] danbri: ivan: one more thing on this issue, purely editorial
[07:21am] jtandy: [so I think that we agree that we use 'base' for minimal validation - and we've opened the door for _extended_ validation]
[07:21am] danbri: … presume gregg will put some note saying something about this, xsd datatypes etc
[07:21am] danbri: ivan: i put a ref to OWL just to clarify that this is not exclusively for xsd
[07:21am] danbri: gkellogg; how would anyone figure out what this _is_ for?
[07:21am] danbri: ivan: let's not get into extra props for that
[07:21am] danbri: gkellogg: ok
[07:22am] danbri: ivan: if you have decided property names, email me and i'll add to diagram

The gist of this to use something other than `@id` in a datatype definition to avoid JSON-LD confusion, perhaps `extension`. Note that if `base` isn't defined, it effectively defaults to `string`.

**Comment by gkellogg** [comment]
@iherman @6a6d74, I've added the `restriction` property on this branch. @JeniT, if you would go over the metadata and syntax updates; I tried to keep it simple.

**Comment by iherman** [comment]
> On 15 May 2015, at 24:24 , Gregg Kellogg notifications@github.com wrote:
> 
> @iherman @6a6d74, I've added the restriction property on this branch. @JeniT, if you would go over the metadata and syntax updates; I tried to keep it simple.

The diagram (on this branch) has been extended to include the restriction property

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by JeniT** [comment]
Sorry, I obviously missed important discussion on the phone. Why can't we use `@id` here? I thought we were wanting to provide a URL identifier for the datatype which is being described by the JSON-LD object, and `@id` seems entirely appropriate for that.

**Comment by gkellogg** [comment]
@iherman, on Wednesday's call, we tentatively decided to go back to using `@id` for this, rather than `restriction`. (It's in the minutes, which I don't seem to be able to find, perhaps you can look into that).

The idea is that `{"datatype": "string"}` is consistent with an expansion to `{"datatype": {"@id": "xsd:string", "base": "string"}}`, so that `{"datatype": {"@id": "http://example.org/type-restriction", "base": "string"}}` makes sense. We voted on this, but didn't have a quorum (there's a resolution floating around somewhere), and wanted you're opinion before preceding.

**Comment by iherman** [comment]
> On 21 May 2015, at 21:11, Gregg Kellogg notifications@github.com wrote:
> 
> @iherman, on Wednesday's call, we tentatively decided to go back to using @id for this, rather than restriction. (It's in the minutes, which I don't seem to be able to find, perhaps you can look into that).
> 
> The idea is that {"datatype": "string"} is consistent with an expansion to {"datatype": {"@id": "xsd:string", "base": "string"}}, so that {"datatype": {"@id": "http://example.org/type-restriction", "base": "string"}} makes sense. We voted on this, but didn't have a quorum (there's a resolution floating around somewhere), and wanted you're opinion before preceding.

That is indeed consistent. So ok for @id.

I am not sure when I will change the diagram, but will have it on the week end at the latest.

Ivan

**Comment by iherman** [comment]
The core metadata spec has been updated via #565. Next step is to add a reference to the @id value into the RDF conversion....

/Cc @6a6d74

**Comment by JeniT** [comment]
The text that's been merged via #565 doesn't look right to me. It talks about the id of the datatype identifying "further restrictions of the datatype" rather than being the identifier for the datatype itself, which is just confusing. We also discussed on the call turning the `datatype` property into an object property rather than an atomic property, and adding some exceptions around the retrieval of known (built-in) datatypes.

Should I raise a separate issue about this or do another pull request to propose fixes?

**Comment by gkellogg** [comment]
No, I'll take care of it. I missed these details.

**Comment by iherman** [comment]
Does it mean I will have to change the diagram for the property type?

/Cc @gkellogg 

---

Ivan Herman
Tel:+31 641044153
http://www.ivan-herman.net

(Written on mobile, sorry for brevity and misspellings...)

**Comment by gkellogg** [comment]
We can probably just revert to the previous version of the diagram, but I'll leave that to you, @iherman.

**Comment by iherman** [comment]
The previous version used 'restriction', that is now '@id'. However, per the comment of @JeniT, the 'datatype' property's color must be changed. Is that correct?

Ivan

---

Ivan Herman
Tel:+31 641044153
http://www.ivan-herman.net

(Written on mobile, sorry for brevity and misspellings...)

**Comment by gkellogg** [comment]
Ahh, yes, that's true.

**Comment by iherman** [comment]
@gkellogg: before I create problems: in which branch are the latest editings for this issue? The main branch still denotes 'datatype' as an atomic property in section 5.7....

**Comment by gkellogg** [comment]
Latest work is on issue-543-datatypes branch, which you can go ahead and use. @JeniT has more issues with it, but this shouldn't affect the diagram.

**Comment by iherman** [comment]
Thanks. I have committed the diagram change.

/Cc @gkellogg @JeniT

> On 24 May 2015, at 08:56 , Gregg Kellogg notifications@github.com wrote:
> 
> Latest work is on issue-543-datatypes branch, which you can go ahead and use. @JeniT has more issues with it, but this shouldn't affect the diagram.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
Replicating commit note here:

@JeniT, still not ideal; it probably should be treated as a document-relative URL, like other object properties. However, this means that using "anySimpleType" would expand to http://example.org/anySimpleType (presuming the appropriate document base), where this should otherwise be detected as an error. However, if you define a local datatype in the same directory (e.g., my-datatype.json), and load that to get a Datatype definition, it would be appropriate to use document relative. But our rules say that if this does not result in a JSON Object then it is ignored, and we create a new datatype with @id based on that expanded URL and base: string. This could allow http://example.org/anySimpleType (or any other XSD type we don't specifically use) when this would otherwise have either been ignored or been an error.

This stems from allowing `datatype` to take any string, which would be interpreted as a URL. The text I added says "absolute URL" which avoids this, but narrows the case when we want to use a local datatype definition in another file. Consider:

```
{"datatype": "string"} #=> {"datatype": {"@id": "xsd:string", "base": "string"}}
{"datatype": "local-datatype.jsonld"} #=> {"datatype": {"base": "my-local-datatype-base", ...}}
{"datatype": "http://example.org/dt.owl"} #=> {"datatype": {"@id": "http://example.org/dt.owl"}}
{"datatype": "anySimpleString"} #=> {"datatype": {"@id": "http://example.org/anySimpleString"}}
```

We presumably want to avoid the last use case but allow the proceeding. Sorry that I don't think we're quite done with this one.

/cc @iherman

**Comment by iherman** [comment]
We are talking about an edge case here, aren't we? The experience with deployed RDF out there may be significant: although the RDF spec indeed allows the usage of any URI to denote a datatype, I have hardly seen any other usage than the predefined XSD-s. We already offer more than that, for validation purposes (with the various restrictions).

As a consequence, I think we should keep it simple: `@datatype` can either be an absolute URI, or a simple string denoting the set of predefined datatypes. (If a person is savy enough to define a new datatype, that person would have no problem putting an absolute URI there.)

(This comment is also valid for #583.)

**Comment by gkellogg** [comment]
As it's an object property, the other case is when it references another file containing a Datatype definition. If this was in error, it wouldn't be flagged as a warning, given the existing wording.

**Comment by 6a6d74** [comment]
I have now merged in the changes to csv2\* conversion docs (PR #568). 

@JeniT, @gkellogg, @iherman ... can this issue now be closed?

**Comment by iherman** [comment]
From my point of view (as the guardian of the diagrams:-): yes.

**Comment by iherman** [comment]
https://github.com/w3c/csvw/issues/543#issuecomment-108651635 is still pending...

Cc: @gkellogg @JeniT

**Comment by JeniT** [comment]
I'm a bit confused by https://github.com/w3c/csvw/issues/543#issuecomment-108651635 to be honest. Maybe it helps to go through the examples one by one.

```
{"datatype": "string"}
```

should normalize to

```
{"datatype": {"@id": "http://www.w3.org/2001/XMLSchema#string" }}
```

No additional properties should be added; implementations should recognise `http://www.w3.org/2001/XMLSchema#string` as meaning a string. (Setting both the `@id` and the `base` to `string` makes no sense: you can't have a datatype based on it itself.

---

```
{"datatype": "local-datatype.jsonld"}
```

Assuming that `http://example.org/local-datatype.jsonld` resolves to a JSON file that looks like:

```
{ "@id": "my-local-datatype", ... }
```

this should normalize to:

```
{"datatype": { "@id": "http://example.org/my-local-datatype", ... }}
```

with the properties from the referenced datatype file. The `@id` in the file should be resolved against the base URL of `http://example.org/local-datatype.jsonld` (which could be overwritten by a `@base` property in the `@context`).

---

```
{"datatype": "http://example.org/dt.owl"}
```

Assuming that `http://example.org/dt.owl` doesn't resolve to a JSON file, this should normalize to:

```
{"datatype": {"@id": "http://example.org/dt.owl"}}
```

(with the assumed base type of `string`, ie with no further validation).

---

```
{"datatype": "anySimpleString"}
```

Given `anySimpleString` isn't a recognised type, it should be taken as a relative URL in the same way as the previous examples. Assuming there's no such file you eventually end up with:

```
{"datatype": {"@id": "http://example.org/anySimpleString"}}
```

I don't think we should recognise any old XML Schema term, just the datatypes that we've said we will recognise.

**Comment by iherman** [comment]
Actually, now that you say there may indeed be problems...

In the current text (in this branch) it says, for `datatype`:

> An atomic property that contains either a single string that is the main datatype of the values of the cell or a datatype description object. If the value of this property is a string, it MUST be one of the built-in datatypes defined in [...]; if it is an object then it describes a more specialised datatype.

That means, actually, that this is not in line with _any_ of the examples, except for the very first one. It seems to suggest that externally defined datatypes should be defined _only_ through the `@id` value of a datatype description object.

That being said, the comments in this issue seem to indicate that a `datatype` should be considered as an object property (and the diagram has been changed accordingly!), though this is not yet reflected in the spec. However, an object property means that the value is either a URL (relative or absolute) or an object (in our case, a datatype description object). Furthermore, for all the other cases we have, the URL in an object property is used to access a json file, replacing the object. If that is correct than
- `"datatype" : "string"`` would be expanded to a relative URL, which is not what we want
- `"datatype" : "http://example.org/dt.owl"` would be rejected because it is not a JSON file

Taking into account that, I believe, the use case for externally defined datatypes seem to be few and far between, I would be fine keeping with the original datatype definition and use `@id` only. But that seems to require a change of a telco resolution. Otherwise, `datatype` is not a pure object property but should have a special interpretation; can be done, not pretty...

**Comment by JeniT** [comment]
We discussed, reverting to `datatype` being an atomic property, which can be a string value of a known datatype or an object whose `@id` can be anything.
